File: ui/src/app/campaigns/dto/campaign-stats-dto.ts
import {GroupProgressDto} from "./group-progress-dto";

export interface CampaignStatsDto {
  groupMembers?: [number];
  groupMembersReplied?: [number];
  membersContacted?: number;
  membersContactedButNotReplied?: number;
  membersReplied?: number;
  successRate?: number;
  allMembers?: number;
  progress?: number;
  duration?: string;
  stateDescription: string;

  // This is a placeholder to be calculated by the frontend component.
  // It is not filled-in by the backend.
  groupProgress: GroupProgressDto[];
}

File: esthesis-core-backend/services/srv-campaign/srv-campaign-client/src/main/java/esthesis/service/campaign/dto/CampaignStatsDTO.java
package esthesis.service.campaign.dto;

import esthesis.core.common.AppConstants;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

/**
 * A representation of campaign statistics.
 */
@Data
@NoArgsConstructor
@Accessors(chain = true)
public class CampaignStatsDTO {

	private List<Long> groupMembers;
	private List<Long> groupMembersReplied;
	private long allMembers;
	private long membersContacted;
	private long membersContactedButNotReplied;
	private long membersReplied;
	private BigDecimal successRate;
	private BigDecimal progress;
	private String duration;
	private String stateDescription;
	private AppConstants.Campaign.State state;
	private Instant createdOn;
	private Instant startedOn;
	private Instant terminatedOn;
}
